src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -61,55 +61,49 @@ impl ScalarFunc {
         Ok(Some(input.len() as i64))
     }
 
+    #[inline]
+    fn locate(s: &str, p: &str, pos: usize) -> i64 {
+        s.find(p)
+            .map(|i| (pos + s[..i].chars().count()) as i64)
+            .unwrap_or(0)
+    }
+
     #[inline]
     pub fn locate_2_args(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         let substr = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
         let s = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
-        if substr.is_empty() {
-            return Ok(Some(1));
-        }
-        if s.is_empty() {
-            return Ok(Some(0));
-        }
-        let substr = substr.to_lowercase();
-        let s = s.to_lowercase();
-        let result = match s.find(&substr) {
-            None => 0,
-            Some(i) => 1 + s[..i].chars().count() as i64,
-        };
-        Ok(Some(result))
+        Ok(Some(Self::locate(
+            &s.to_lowercase(),
+            &substr.to_lowercase(),
+            1,
+        )))
     }
 
     #[inline]
     pub fn locate_3_args(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         let substr = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
         let s = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
-        let mut pos = try_opt!(self.children[2].eval_int(ctx, row));
-
-        pos -= 1; // transfer 1-based argument to 0-based real index
-        if pos < 0 {
-            return Ok(Some(0));
-        }
-        let substr = substr.to_lowercase();
-        let s = s.to_lowercase();
-        if pos > (s.chars().count() as i64 - substr.chars().count() as i64) {
-            return Ok(Some(0));
-        }
-        if substr.is_empty() {
-            return Ok(Some(pos + 1));
-        }
-
-        let slice = s
-            .char_indices()
-            .nth(pos as usize)
-            .map(|(offset, _)| &s[offset..]);
+        let pos = try_opt!(self.children[2].eval_int(ctx, row));
+        Ok(if pos < 1 {
+            Some(0)
+        } else {
+            let pos = pos as usize;
+            s.char_indices()
+                .map(|(i, _)| i)
+                .chain([s.len()].iter().cloned())
```

Comment:
I'm worried about the performance here. I guess the compiler is not smart enough to optimize the clone away. Could you write a benchmark?

Reply:
Why not use `std::iter::once()`

Reply:
@kennytm thx!

Reply:
> I'm worried about the performance here. I guess the compiler is not smart enough to optimize the clone away. Could you write a benchmark?

maybe we can just chain a `Once` to the tail ?
```
pub fn once<T>(value: T) -> Once<T> // no clone here!
```